### PR TITLE
feat: sdk revamp

### PR DESCRIPTION
**Pull request by rsimari**

### CODE REVIEW

**Review by s-ashwinkumar — COMMENTED** [review]
- nit: ApplicationPermissions#setPaging - not used, remove
- nit: Remove ApplicationPermissions#setPaging
- nit: BatchResponse#toCamelCase and ApiClient#toCamelCase is the same can we move the if cotains '_' logic into caseUtils.toCamelCase and call that directly on the lambda ? is that even possible ?
- nit: I see the whole `Request.Builder()...` code invoked everywhere, maybe consider a util method for that which takes in the url and type of auth OR the value for auth header. We dont have to repeat things like useragent etc.

File: src/main/java/com/smartcar/sdk/data/Meta.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+    public String getRequestId() { return this.requestId; }
+
+    public Date getDataAge() throws SmartcarException {
+        try {
+            DateFormat format = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
+            return format.parse(this.dataAge);
+        } catch (ParseException ex) {
+            throw new SmartcarException.Builder().type("SDK_ERROR").description(ex.getMessage()).build();
+        }
```
[the remark this answers is not in the document]
Comment (rsimari) [reply src/main/java/com/smartcar/sdk/data/Meta.java:29]: we are forced here. if not we need the method signature to have a `throws ParseException`

File: src/main/java/com/smartcar/sdk/SmartcarCompatibilityRequest.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        }
+
+        public Builder clientSecret(String clientSecret) {
+            this.clientSecret = clientSecret;
+            return this;
+        }
+
+        public SmartcarCompatibilityRequest build() throws SmartcarException {
+            if (this.clientId == null) {
+                throw new SmartcarException.Builder()
```
[the remark this answers is not in the document]
Comment (rsimari) [reply src/main/java/com/smartcar/sdk/SmartcarCompatibilityRequest.java:76]: switched to `Exception` makes sense to me

File: src/main/java/com/smartcar/sdk/Smartcar.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+     * @return
+     */
+    static String getApiUrl() {
+        return getApiOrigin() + "/v" + Smartcar.API_VERSION;
+    }
+
+    static String getApiOrigin() {
+        String apiOrigin = System.getenv("SMARTCAR_API_ORIGIN");
+        if (apiOrigin == null) {
+            return "https://api.smartcar.com";
```
[the remark this answers is not in the document]
Comment (rsimari) [reply src/main/java/com/smartcar/sdk/Smartcar.java:41]: for versioning? `getApiUrl` uses version and `getApiOrigin` doesnt so we can use the the compatibility version separately which might be different than `SMARTCAR_API_ORIGIN`. the `SDK_VERSION` is the actual package version number not the api version

File: src/main/java/com/smartcar/sdk/ApiClient.java [hunk 4]
```diff
@@ -1,28 +1,28 @@
...
+import okhttp3.Headers;
 import org.apache.commons.text.CaseUtils;
 
 /** Provides the core functionality for API client objects. */
 abstract class ApiClient {
   public static final MediaType JSON = MediaType.parse("application/json; charset=utf-8");
-  private static final String SDK_VERSION = ApiClient.getSdkVersion();
-  private static String API_VERSION = "1.0";
-  protected static final String URL_API = "https://api.smartcar.com";
+
```
[the remark this answers is not in the document]
Comment (rsimari) [reply src/main/java/com/smartcar/sdk/ApiClient.java:16]: `System.getenv` will only throw if the input is `null` or there is some other permissions issue with the process and its env. So it returns `null` for any values that aren't found.

[the remark this answers is not in the document]
Comment (rsimari) [reply src/main/java/com/smartcar/sdk/ApiClient.java:16]: I added in the checks for redirectUri, clientId, clientSecret in `AuthClient.Builer.build`

File: src/main/java/com/smartcar/sdk/ApiClient.java [hunk 5]
```diff
@@ -109,32 +73,33 @@ protected static Response execute(Request request) throws SmartcarException {
...
-    String body = null;
+    T data;
+    Meta meta;
+    String bodyString = null;
 
     try {
-      body = response.body().string();
-    } catch (IOException ex) {
-      throw new SmartcarException(ex.getMessage());
-    }
```
[the remark this answers is not in the document]
Comment (rsimari) [reply src/main/java/com/smartcar/sdk/ApiClient.java:121]: i made a test for this case. added the correct handling

File: src/main/java/com/smartcar/sdk/Smartcar.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (s-ashwinkumar) [reply src/main/java/com/smartcar/sdk/Smartcar.java:43]: Wrong comment by me....we are deleting the `API_ORIGIN` since its only in 2 places

**Review by s-ashwinkumar — APPROVED** [review]

**Pull request merged**